src/core/common/trickle_timer.cpp

```diff
@@ -53,6 +53,78 @@ TrickleTimer::TrickleTimer(Instance &aInstance, Handler aHandler)
 {
 }
 
+TimeMilli TrickleTimer::GetLastIntervalStart(void)
+{
+    TimeMilli startTime = TimerMilli::GetFireTime();
+
+    if (mMode == kModePlainTimer)
+    {
+        startTime -= mInterval;
+        ExitNow();
+    }
+
+    switch (mPhase)
+    {
+    case kBeforeRandomTime:
+        startTime -= mTimeInInterval;
+        break;
+
+    case kAfterRandomTime:
+        startTime -= mInterval;
+        break;
+    }
+
+exit:
+    return startTime;
+}
+
+void TrickleTimer::SetIntervalMax(uint32_t aIntervalMax)
+{
+    OT_ASSERT(aIntervalMax >= mIntervalMin);
+
+    VerifyOrExit(aIntervalMax != mIntervalMax);
+
+    mIntervalMax = aIntervalMax;
+
+    if (aIntervalMax < mInterval)
+    {
+        // the new interval is smaller than the running interval, so re-configure the timer to
+        // fire at the maximum value
+
+        TimeMilli newFireTime = GetLastIntervalStart();
+        newFireTime += aIntervalMax;
+
+        // In plain mode, just fire at the sooner time and
+        // the handler will take care of everything else.
+        if (mMode == kModePlainTimer)
+        {
+            TimerMilli::FireAt(newFireTime);
+            ExitNow();
+        }
+
+        // In trickle mode, there are two different phases
+        // when in kBeforeRandomTime:
+        //   - aIntervalMax < mTimeInInterval - need to set mInterval = aIntervalMax,
+        //     mTimeInInteval = aIntevalMax & re-schedule timer to fire at
+        //     aIntervalMax (from start of interval). The HandleTimer function will
+        //     take care of the transition through kAfterRandomTime
+        //   - aIntervalMax >= mTimeInInterval - need to set mInterval = aIntervalMax
+        // when in kAfterRandomTime, in both cases need to set mInteval = aInterMax & re-schedule
+        //     timer to fire at aIntervalMax (from start of interval)
+
+        mInterval = aIntervalMax;
+        if (mTimeInInterval > aIntervalMax)
+        {
+            mTimeInInterval = aIntervalMax;
+        }
+
+        TimerMilli::FireAt(newFireTime);
```

Comment:
Here is what I would suggest:
- In all cases we do `mInterval = aIntervalMax;`.
- The `if()` is for `i.b` case where we exit (and won't change the timer).
- In all other cases, we `mTimeInInterval = aIntervalMax` (only matters for `1.a` case but harmless to do in other cases), and re-schedule the timer.


```suggestion
        mInterval = aIntervalMax;

        if ((mPhase == kBeforeRandomTime)) && (aIntervalMax >= mTimeInInterval))
        {
        	ExitNow();
        }

        mTimeInInterval = aIntervalMax;
        TimerMilli::FireAt(newFireTime);
```